pallets/parachain-staking/src/lib.rs

```diff
@@ -2768,59 +2788,77 @@ pub mod pallet {
 			inner.try_into().expect("Did not extend vec q.e.d.")
 		}
 
-		fn peaq_reward_mechanism_impl() {
-			let mut reads = Weight::from_parts(0, 1);
-			let mut writes = Weight::from_parts(0, 1);
+		/// Get a unique, inaccessible account id from the `PotId`.
+		pub fn account_id() -> T::AccountId {
+			T::PotId::get().into_account_truncating()
+		}
 
+		/// Handles staking reward payout for previous session for one collator and their delegators
+		fn payout_collator() {
 			let pot = Self::account_id();
-			let issue_number = T::Currency::free_balance(&pot)
-				.checked_sub(&T::Currency::minimum_balance())
-				.unwrap_or_else(Zero::zero);
+			// get payout info for the last round
+			let payout_info = DelayedPayoutInfo::<T>::get();
 
-			let (in_reads, total_staking_in_session) = Self::get_total_collator_staking_num();
-			reads.saturating_add(in_reads);
-
-			// Here we also remove the all collator block after the iteration
-			CollatorBlock::<T>::iter().drain().for_each(|(collator, block_num)| {
-				// Get the delegator's staking number
-				if let Some(state) = CandidatePool::<T>::get(collator.clone()) {
+			// get one author
+			if let Some((author, block_num)) =
+				CollatorBlocks::<T>::iter_prefix(payout_info.round).drain().next()
+			{
+				// get collator's staking info
+				if let Some(state) = AtStake::<T>::take(payout_info.round, author) {
+					// calculate reward for collator from previous round
 					let now_reward = Self::get_collator_reward_per_session(
 						&state,
 						block_num,
-						total_staking_in_session,
-						issue_number,
+						payout_info.total_stake,
+						payout_info.total_issuance,
```

Comment:
Is that possible payout_info.total_issuance number is 0 while the migration happens? If so, then we should consider some error handling here

Reply:
So because we take a snapshot of active collators at runtime upgrade with a storage migration, prepare_staking_rewards will calculate DelayedPayoutInfo based off of that, so we don't expect total_issuance to be 0. 
But this is a good point, i'll introduce better error handling.

Reply:
Cool, if we don't expect the total_issuance to be 0, I think that's totally fine